### PR TITLE
[CBRD-23602] regression: Statement cache to improve performance for frequently repeated queries

### PR DESCRIPTION
**Pull request by byungwook-kim**
http://jira.cubrid.org/browse/CBRD-23602

This a revised patch for #2443 
- patched for fatal bug: cub_server core-dump
- patched for dirty-read.
- patched for the following issues:
  http://jira.cubrid.org/browse/CBRD-23771
  http://jira.cubrid.org/browse/CBRD-23772
  http://jira.cubrid.org/browse/CBRD-23773
  http://jira.cubrid.org/browse/CBRD-23774
  http://jira.cubrid.org/browse/CBRD-23775
  http://jira.cubrid.org/browse/CBRD-23776
  http://jira.cubrid.org/browse/CBRD-23777
  http://jira.cubrid.org/browse/CBRD-23778
  http://jira.cubrid.org/browse/CBRD-23779
  http://jira.cubrid.org/browse/CBRD-23780
  http://jira.cubrid.org/browse/CBRD-23781
  http://jira.cubrid.org/browse/CBRD-23785

### CODE REVIEW

File: src/query/list_file.c [hunk 1]
```diff
@@ -275,7 +275,6 @@ static int qfile_free_list_cache_entry (THREAD_ENTRY * thread_p, void *data, voi
 static int qfile_print_list_cache_entry (THREAD_ENTRY * thread_p, FILE * fp, const void *key, void *data, void *args);
 static void qfile_add_uncommitted_list_cache_entry (int tran_index, QFILE_LIST_CACHE_ENTRY * lent);
 static void qfile_delete_uncommitted_list_cache_entry (int tran_index, QFILE_LIST_CACHE_ENTRY * lent);
-static int qfile_delete_list_cache_entry (THREAD_ENTRY * thread_p, void *data, void *args);
```
Comment (shparkcubrid) [inline src/query/list_file.c:278]: why to delete this?

File: src/query/list_file.c [hunk 2]
```diff
@@ -371,7 +370,7 @@ qfile_list_cache_cleanup (THREAD_ENTRY * thread_p)
   for (candidate_index = 0; candidate_index < bh->element_count; candidate_index++)
     {
       bh_element_at (bh, candidate_index, &candidate);
-      qfile_end_use_of_list_cache_entry (thread_p, candidate.qcache, true);
+      qfile_delete_list_cache_entry (thread_p, candidate.qcache, NULL);
```
Comment (shparkcubrid) [inline src/query/list_file.c:373]: What is the difference between qfile_end_use_of_list_cache_entry() and qfile_delete_list_cache_entry()?
Is it safe to delete immediately?

File: src/storage/file_manager.c [hunk 3]
```diff
@@ -4324,7 +4324,12 @@ file_temp_retire_internal (THREAD_ENTRY * thread_p, const VFID * vfid, bool was_
   else
     {
       entry = file_tempcache_pop_tran_file (thread_p, vfid);
-      assert (entry != NULL);
+      /* should have found it
+         assert (entry != NULL);
+         * but coulld be already removed by result-cache processing
+         * so, disable assertion
+       */
```
Comment (hgryoo) [inline src/storage/file_manager.c:4331]: please revert comment out

File: src/storage/file_manager.c [hunk 4]
```diff
@@ -8546,7 +8551,11 @@ file_temp_preserve (THREAD_ENTRY * thread_p, const VFID * vfid)
   entry = file_tempcache_pop_tran_file (thread_p, vfid);
   if (entry == NULL)
     {
-      assert_release (false);
+      /* should have found it
+         assert_release (false);
+         * but coulld be already removed by result-cache processing
+         * so, disable assertion
+       */
```
Comment (hgryoo) [inline src/storage/file_manager.c:8558]: please revert comment out

File: src/storage/file_manager.c [hunk 5]
```diff
@@ -9118,8 +9127,12 @@ file_tempcache_pop_tran_file (THREAD_ENTRY * thread_p, const VFID * vfid)
       prev_entry = entry;
     }
 
-  /* should have found it */
-  assert_release (false);
+  /* should have found it
+     assert_release (false);
+     * but coulld be already removed by result-cache processing
+     * so, disable assertion
+   */
```
Comment (hgryoo) [inline src/storage/file_manager.c:9134]: please revert comment out

**Review by hgryoo — CHANGES_REQUESTED** [review]
It could be safe that only when result-cache is enabled, ignore the assert.

File: src/query/query_manager.c [hunk 6]
```diff
@@ -2841,7 +2786,7 @@ qmgr_free_temp_file_list (THREAD_ENTRY * thread_p, QMGR_TEMP_FILE * tfile_vfid_p
       fd_ret = NO_ERROR;
       if ((tfile_vfid_p->temp_file_type != FILE_QUERY_AREA || is_error) && !VFID_ISNULL (&tfile_vfid_p->temp_vfid))
 	{
-	  if (was_preserved)
+	  if (was_preserved || tfile_vfid_p->temp_file_type == FILE_QUERY_AREA)
```
Comment (hgryoo) [inline src/query/query_manager.c:2789]: This if condition seems not clear.
there exists "tfile_vfid_p->temp_file_type != FILE_QUERY_AREA" at outer if-statement and tfile_vfid_p->temp_file_type == FILE_QUERY_AREA at inner if-statement.

File: src/storage/file_manager.c [hunk 7]
```diff
@@ -8546,7 +8551,11 @@ file_temp_preserve (THREAD_ENTRY * thread_p, const VFID * vfid)
   entry = file_tempcache_pop_tran_file (thread_p, vfid);
   if (entry == NULL)
     {
-      assert_release (false);
```
Comment (wrlawodms) [inline src/storage/file_manager.c:8549]: I think this assertion says we have to use `preserve` interface to pop out temp files from the transaction list. Do you happen to make a new interface to handle temp files without preserving? Can I know where to pop a temp file from a transaction list and clean up the file?

File: src/storage/file_manager.c [hunk 8]
```diff
@@ -4324,7 +4324,12 @@ file_temp_retire_internal (THREAD_ENTRY * thread_p, const VFID * vfid, bool was_
   else
     {
       entry = file_tempcache_pop_tran_file (thread_p, vfid);
-      assert (entry != NULL);
```
Comment (wrlawodms) [inline src/storage/file_manager.c:4327]: Is there a state in which a temp file is not related to a transaction and not preserved? Why don't you preserve it?

File: src/query/list_file.c (hunk 1 above)
Comment (byungwook-kim) [reply src/query/list_file.c:278]: I changed static to global.

File: src/query/list_file.c (hunk 2 above)
Comment (byungwook-kim) [reply src/query/list_file.c:383]: qfile_end_use_of_list_cache_entry: remove only the related information
qfile_delete_list_cache_entry: remove the result file from cache entry.

File: src/query/query_manager.c (hunk 6 above)
Comment (byungwook-kim) [reply src/query/query_manager.c:2765]: have to check the error case (is_error is true)

File: src/storage/file_manager.c (hunk 7 above)
Comment (byungwook-kim) [reply src/storage/file_manager.c]: I fixed it to revert it.

File: src/storage/file_manager.c (hunk 8 above)
Comment (byungwook-kim) [reply src/storage/file_manager.c:4327]: I fixed it to revert it.

File: src/query/query_manager.c (hunk 6 above)
Comment (byungwook-kim) [reply src/query/query_manager.c]: I fixed it as using preserved flag in QMGR temp_file.

File: src/storage/file_manager.c (hunk 3 above)
Comment (byungwook-kim) [reply src/storage/file_manager.c]: I fixed it to revert it.

File: src/storage/file_manager.c (hunk 4 above)
Comment (byungwook-kim) [reply src/storage/file_manager.c]: I fixed it to revert it.

File: src/storage/file_manager.c (hunk 5 above)
Comment (byungwook-kim) [reply src/storage/file_manager.c]: I fixed it to revert it.

File: src/query/list_file.c (hunk 1 above)
Comment (byungwook-kim) [reply src/query/list_file.c:278]: I reverted to static function.

File: src/storage/file_manager.c [hunk 9]
```diff
@@ -4323,12 +4323,8 @@ file_temp_retire_internal (THREAD_ENTRY * thread_p, const VFID * vfid, bool was_
     }
   else
     {
+      assert (entry != NULL);
       entry = file_tempcache_pop_tran_file (thread_p, vfid);
```
Comment (hgryoo) [inline src/storage/file_manager.c]: the order of code is changed

Comment (byungwook-kim) [reply src/storage/file_manager.c]: I fixed it. thanks.

File: src/query/list_file.c [hunk 10]
```diff
@@ -319,8 +319,18 @@ qfile_list_cache_cleanup (THREAD_ENTRY * thread_p)
   int candidate_index, i, n;
 
   struct timeval current_time;
-  int cleanup_count = prm_get_integer_value (PRM_ID_LIST_MAX_QUERY_CACHE_ENTRIES) * 0.8;
-  int cleanup_pages = prm_get_integer_value (PRM_ID_LIST_MAX_QUERY_CACHE_PAGES) * 0.8;
+  int cleanup_count = prm_get_integer_value (PRM_ID_LIST_MAX_QUERY_CACHE_ENTRIES) * 10 / 8;
+  int cleanup_pages = prm_get_integer_value (PRM_ID_LIST_MAX_QUERY_CACHE_PAGES) * 10 / 8;
```
Comment (shparkcubrid) [inline src/query/list_file.c:323]: I think 10 / 8 ==> 8 /10

File: src/query/list_file.c [hunk 11]
```diff
@@ -5527,7 +5537,6 @@ qfile_clear_uncommited_list_cache_entry (THREAD_ENTRY * thread_p, int tran_index
   lent = qfile_List_cache.tran_list[tran_index];
   while (lent)
     {
-      lent->uncommitted_marker = false;
```
Comment (shparkcubrid) [inline src/query/list_file.c:5530]: Is it okay if uncommitted_marker is not initialized to false?

File: src/query/query_manager.c [hunk 12]
```diff
@@ -2186,26 +2134,18 @@ qmgr_clear_trans_wakeup (THREAD_ENTRY * thread_p, int tran_index, bool is_tran_d
...
-      if (!query_p->is_holdable || is_abort)
+      if (query_p->xasl_ent != NULL && query_p->list_ent != NULL)
 	{
-	  if (query_p->xasl_ent != NULL && query_p->list_ent != NULL)
-	    {
-	      (void) qfile_end_use_of_list_cache_entry (thread_p, query_p->list_ent, false);
-	    }
-	  query_p->query_status = QUERY_CLOSED;
+	  (void) qfile_end_use_of_list_cache_entry (thread_p, query_p->list_ent, false);
+	  query_p->xasl_ent = NULL;
```
Comment (hgryoo) [inline src/query/query_manager.c:2140]: where query_p->xasl_ent is deallocated?

File: src/query/query_manager.c [hunk 13]
```diff
@@ -2841,7 +2785,7 @@ qmgr_free_temp_file_list (THREAD_ENTRY * thread_p, QMGR_TEMP_FILE * tfile_vfid_p
       fd_ret = NO_ERROR;
       if ((tfile_vfid_p->temp_file_type != FILE_QUERY_AREA || is_error) && !VFID_ISNULL (&tfile_vfid_p->temp_vfid))
 	{
-	  if (was_preserved)
```
Comment (hgryoo) [inline src/query/query_manager.c:2844]: what was the context of was_preserved?
A `preserved` variable in `QMGR_TEMP_FILE` is the same meaning?

File: src/query/query_manager.c (hunk 12 above)
Comment (byungwook-kim) [reply src/query/query_manager.c:2140]: let me check of it, it looks something wrong.

File: src/query/query_manager.c (hunk 13 above)
Comment (byungwook-kim) [reply src/query/query_manager.c:2844]: same meaning because that the QMGR's preserved flag is only set at 'create result file'

File: src/query/list_file.c (hunk 10 above)
Comment (byungwook-kim) [reply src/query/list_file.c:323]: I fixed it. thanks.

File: src/query/list_file.c (hunk 11 above)
Comment (byungwook-kim) [reply src/query/list_file.c:5530]: the following functions are not used:
  qfile_add_uncommitted_list_cache_entry
  qfile_clear_uncommited_list_cache_entry
  qfile_delete_uncommited_list_cache_entry
for transaction contention problem between cached result.

**Review by shparkcubrid — APPROVED** [review]

File: src/query/list_file.c [hunk 14]
```diff
@@ -6760,3 +6770,11 @@ qfile_update_qlist_count (THREAD_ENTRY * thread_p, const QFILE_LIST_ID * list_p,
     }
 #endif // SERVER_MODE
 }
+
+int
+qfile_get_list_cache_number_of_entries (THREAD_ENTRY * thread_p, int ht_no)
```
Comment (hgryoo) [inline src/query/list_file.c:6775]: Because this function doesn't access thread-specific data, THREAD_ENTRY  is not needed

File: src/query/list_file.c [hunk 15]
```diff
@@ -5188,13 +5198,16 @@ qfile_clear_list_cache (THREAD_ENTRY * thread_p, int list_ht_no, bool release)
       /* unhappy condition; if this happens, memory leak will occurrs */
       er_log_debug (ARG_FILE_LINE, "ls_clear_list_cache: failed to delete all entries\n");
     }
-  (void) mht_clear (qfile_List_cache.list_hts[list_ht_no], NULL, NULL);
 
-  /* release assigned memory hash table */
-  if (release)
+  if (qfile_List_cache.list_hts[list_ht_no]->nentries == 0)
```
Comment (hgryoo) [inline src/query/list_file.c:5202]: intended to use `qfile_get_list_cache_number_of_entries ()`?

File: src/query/query_manager.c (hunk 13 above)
Comment (hgryoo) [reply src/query/query_manager.c:2888]: The result file is whether preserved or not was managed in `query entry`. But now, each result file sets preserved itself. and `query_entry->is_preserved` seems not used anymore. In terms of this changing of context, Is it safe? and `query_entry->is_preserved` is still needed?

Comment (hgryoo) [reply src/query/query_manager.c:2888]: We discussed that `qmgr_query_entry.is_preserved` becomes redundant variable now. It would better to remove it.

File: src/query/list_file.c (hunk 15 above)
Comment (byungwook-kim) [reply src/query/list_file.c:5202]: I fixed it to use qfile_get_list_cache_number_of_entries.

File: src/query/list_file.c (hunk 14 above)
Comment (byungwook-kim) [reply src/query/list_file.c]: I fixed it to remove the argument.

**Review by hgryoo — APPROVED** [review]
